src/aspire/denoise/covar2d_ctf.py

```diff
@@ -16,31 +28,142 @@ def get_mean_ctf(self, coeffs, ctf_fb, ctf_idx):
+            mean_coeff_k = blk_diag_apply(ctf_fb_k, mean_coeff)
+            covar_coeff_k = self.get_covar(coeff_k, mean_coeff_k)
+
+            b_coeff = blk_diag_add(b_coeff, blk_diag_mult(ctf_fb_k_t,
+                blk_diag_mult(covar_coeff_k, blk_diag_mult(ctf_fb_k, weight))))
+
+            b_noise = blk_diag_add(b_noise, blk_diag_mult(weight,
+                blk_diag_mult(ctf_fb_k_t, ctf_fb_k)))
+
+            A[k] = blk_diag_mult(ctf_fb_k_t, blk_diag_mult(ctf_fb_k, np.sqrt(weight)))
+
+            M = blk_diag_add(M, A[k])
+
+        if covar_est_opt['shrinker'] == 'none':
+            b = blk_diag_add(b_coeff, blk_diag_mult(-noise_var, b_noise))
+        else:
+            b = self.shrink_covar_backward(b_coeff, b_noise, np.size(coeffs, 1),
+                                           noise_var, covar_est_opt['shrinker'])
+
+        cg_opt = covar_est_opt
+        covar_coeff = blk_diag_zeros(block_partition, dtype=self.as_type)
+
+        def precond_fun(S, x):
+            p = np.size(S, 0)
+            x = m_reshape(x, p*np.ones((1, 2)))
+            y = S @ x @ S
+            return y
+
+        def apply(A, x):
+            p = np.size(A[0], 0)
+            x = m_reshape(x, p*np.ones((1, 2)))
+            y = np.zeros_like(x)
+            for k in range(0, len(A)):
+                    y = y + A[k] @ x @ A[k].T
+            return y
+
+        for ell in range(0, len(b)):
+            A_ell = []
+            for k in range(0, len(A)):
+                A_ell.append(A[k][ell])
+            b_ell = b[ell]
+            S = inv(M[ell])
+            cg_opt["preconditioner"] = lambda x: precond_fun(S, x)
+            covar_coeff[ell], _, _ = conj_grad(lambda x: apply(A_ell, x), b_ell, cg_opt)
+
+        return covar_coeff
+
+    def shrink_covar_backward(self, b, b_noise, n, noise_var, shrinker):
+        b_out = b
+        for ell in range(0, b.size()):
+            b_ell = b[ell]
+            p = np.size(b_ell, 1)
+            S = sqrtm(b_noise[ell])
+            # from Matlab b_ell = S \ b_ell /S
+            b_ell = np.divide(solve(S,b_ell), S)
+            b_ell = shrink_covar(b_ell, noise_var, p/n, shrinker)
+            b_ell = S @ b_ell @ S
+            b_out[ell] = b_ell
+        return b_out
+
+    def get_wiener_ctf(self, coeffs, ctf_fb, ctf_idx, mean_coeff=None, covar_coeff=None, noise_var=1):
+        if mean_coeff is None:
+            mean_coeff = self.get_mean_ctf(coeffs, ctf_fb, ctf_idx)
+        if covar_coeff is None:
+            covar_coeff = self.get_covar_ctf(coeffs, ctf_fb, ctf_idx, mean_coeff, noise_var=noise_var)
+
+        blk_partition = blk_diag_partition(ctf_fb[0])
+
+        noise_covar_coeff = blk_diag_mult(noise_var, blk_diag_eye(blk_partition, dtype=self.as_type))
+
+        coeffs_est = np.zeros_like(coeffs, dtype=self.as_type)
+
+        for k in np.unique(ctf_idx[:]):
+            coeff_k = coeffs[:, ctf_idx == k]
+            ctf_fb_k = ctf_fb[k]
+            print("ctf_fb_k", ctf_fb_k)
```

Comment:
Don't print stuff. Use logger. Use it generously.

Reply:
Those have been removed.